Platform/OpenNtfsDxe/Index.c

```diff
@@ -0,0 +1,936 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+extern UINTN  mFileRecordSize;
+extern UINTN  mIndexRecordSize;
+STATIC UINT64 mBufferSize;
+
+STATIC
+VOID
+FreeNode (
+  IN NTFS_FILE  *Node    OPTIONAL,
+  IN FSHELP_CTX *Context
+  )
+{
+  ASSERT (Context != NULL);
+
+  if (Node != Context->RootNode) {
+    if (Node != NULL) {
+      FreeFile (Node);
+      FreePool (Node);
+    }
+  }
+}
+
+STATIC
+VOID
+PopElement (
+  IN FSHELP_CTX *Context
+  )
+{
+  STACK_ELEMENT *Element;
+
+  ASSERT (Context != NULL);
+
+  Element = Context->CurrentNode;
+  Context->CurrentNode = Element->Parent;
+  FreeNode (Element->Node, Context);
+  FreePool (Element);
```

Comment:
Sorry if I did not check the context, but is it necessary to check for non-NULL for Element here?